ntp-client/src/Network/NTP/Client.hs

```diff
@@ -1,318 +1,85 @@
-    handleIOException addressFamily e = do
-        traceWith tracer $ NtpTraceReceiveLoopHandleIOException e
-        threadDelay 5000000
-        udpLocalAddresses >>= createAndBindSock tracer addressFamily >>= \case
-            Nothing   -> do
-                traceWith tracer NtpTraceReceiveLoopException
---                logWarning "recreating of sockets failed (retrying)"
-                handleIOException addressFamily e
-            Just sock -> do
-                atomically $ modifyTVar' (ncSockets cli) (\s -> s <> sock)
-                case sock of
-                    This (Last sock_)
-                        -> loop addressFamily $ runWithAddrFamily sock_
-                    That (Last sock_)
-                        -> loop addressFamily $ runWithAddrFamily sock_
-                    These _ _
-                        -> error "NtpClient: startReceive: impossible"
-
-    -- Compute the clock offset based on current time and record it in the NTP
-    -- client state.   A packet will be disgarded if it came after
-    -- @'ntpResponseTimeout'@.
-    handleNtpPacket
-        :: NtpPacket
-        -> IO ()
-    handleNtpPacket packet = do
-        traceWith tracer NtpTraceReceiveLoopPacketReceived -- packet
-        clockOffset (ntpResponseTimeout $ ncSettings cli) packet >>= \case
-            Nothing -> traceWith tracer NtpTraceReceiveLoopLatePacket
-            Just offset -> do
-                traceWith tracer $ NtpTraceReceiveLoopPacketDeltaTime $ getNtpOffset offset
-                atomically $ modifyTVar' (ncState cli) ( offset : )
-
--- |
--- Spawn NTP client which will send request to NTP servers every @'ntpPollDelay'@
--- and will listen for responses.  The @'ncStatus'@ will be updated every
--- @'ntpPollDelay'@ with the most recent value.  It should be run in a separate
--- thread, since it will block infinitely.
-spawnNtpClient :: Tracer IO NtpTrace -> NtpClientSettings -> TVar NtpStatus -> IO ()
-spawnNtpClient tracer settings ncStatus = do
-    traceWith tracer NtpTraceSpawnNtpClientStarting
-    bracket (mkSockets tracer settings) closeSockets $ \sock -> do
-        cli <- mkNtpClient settings ncStatus sock
-
-        addrs <- resolve
-        -- TODO
-        -- we should start listening for requests when we send something, since
-        -- we're not expecting anything to come unless we send something.  This
-        -- way we could simplify the client and remove `ncState` mutable cell.
-        receiveLoop tracer cli
-            `concurrently_` sendLoop tracer cli addrs
-            `concurrently_` traceWith tracer NtpTraceSpawnNtpClientStarted
-
+    ntpStatus <- newTVarIO NtpSyncPending
+    withAsync (ntpClientThread tracer ntpSettings ntpStatus) $ \tid -> do
+        let client = NtpClient
+              { ntpGetStatus = readTVar ntpStatus
+              , ntpTriggerUpdate = do
+                  traceWith tracer NtpTraceTriggerUpdate
+                  atomically $ writeTVar ntpStatus NtpSyncPending
```

Comment:
^ I don't think the wallet will need this method. `ntpGetStatus` and `ntpQueryBlocking` is all what they need.  You could also expose the `TVar`. The wallet already has logic which is using it.

Reply:
OK I removed the `ntpTriggerUpdate`.
We can put the TVar in the API if the wallet really needs it, but I think it is an implementation detail, I would rather not put it in the API because it is not explicit how to use it.
I using a single TVar for reporting the result and forcing an update looks like a hack to me.
(though I admit that it gives a concise implementation)

Reply:
As you feel. Exposing `TVar` gives more power to the user.  My only point was that we don't need `ntpTriggerUpdate` when we expose `ntpQueryBlocking`.